single_kernel_mongo/managers/mongodb_operator.py

```diff
@@ -0,0 +1,875 @@
+            return
+
+        # remove any IPs that are no longer juju hosts & update app data.
+        self.update_hosts()
+        # Add in any new IPs to the replica set. Relation handlers require a reference to
+        # a unit.
+        self.on_relation_changed()
+
+        # make sure all nodes in the replica set have the same priority for re-election. This is
+        # necessary in the case that pre-upgrade hook fails to reset the priority of election for
+        # cluster nodes.
+        self.mongo_manager.set_election_priority(priority=1)
+
+    def update_hosts(self) -> None:
+        """Update the replica set hosts and remove any unremoved replica from the config."""
+        if not self.state.db_initialised:
+            return
+        self.mongo_manager.process_unremoved_units()
+        self.update_related_hosts()
+
+    def update_related_hosts(self) -> None:
+        """Update the app relations that need to be made aware of the new set of hosts."""
+        if self.state.is_role(MongoDBRoles.REPLICATION):
+            for relation in self.state.client_relations:
+                self.mongo_manager.update_app_relation_data(relation)
+            return
+
+        if self.state.is_role(MongoDBRoles.CONFIG_SERVER):
+            # Update the mongos host in the sharded deployment
+            self.config_server_manager.update_mongos_hosts()
+            # Update the config server DB URI on the remote mongos
+            self.cluster_manager.update_config_server_db()
+
+    def open_ports(self) -> None:
+        """Open ports on the workload.
+
+        VM-only.
+        """
+        if self.substrate != Substrates.VM:
+            return
+        ports = [MongoPorts.MONGODB_PORT]
+        if self.state.is_role(MongoDBRoles.CONFIG_SERVER):
+            ports.append(MongoPorts.MONGOS_PORT)
+
+        try:
+            for port in ports:
+                self.workload.exec(["open-port", f"{port}/TCP"])
+        except WorkloadExecError as e:
+            logger.exception(f"Failed to open port: {e}")
+            raise
+
+    def _set_os_config(self) -> None:
+        """Sets sysctl config for mongodb."""
+        try:
+            self.sysctl_config.configure(OS_REQUIREMENTS)
+        except (sysctl.ApplyError, sysctl.ValidationError, sysctl.CommandError) as e:
+            # we allow events to continue in the case that we are not able to correctly configure
+            # sysctl config, since we can  still run the workload with wrong sysctl parameters
+            # even if it is not optimal.
+            logger.error(f"Error setting values on sysctl: {e.message}")
+            # containers share the kernel with the host system, and some sysctl parameters are
+            # set at kernel level.
+            logger.warning("sysctl params cannot be set. Is the machine running on a container?")
+
+    @property
+    def primary(self) -> str | None:
```

Comment:
can we change the name of this to `primary_unit_name` ( i know this was copied directly from the original) but the function has always been a pain to use, since it had to be traced down and be determined what it actually returns and this might be a good opportunity to fix it. If you believe it is out of scope though I respect that